File: Benchmarks/MergeSort/C++/optimized.cpp
#include <cstring>
#include <fstream>
#include <string>
#include <thread>
#include <vector>

using namespace std;

const long int CONCURRENT_THRESHOLD = 10000;

void merge_sort_parallel(const char **start, const char **end);
void merge_sort_concurrent(const char **start, const char **end);

void merge(const char **target, const char **a_start, const char **a_end,
           const char **b_start, const char **b_end) {
  size_t i = 0;

  const char **a = a_start;
  const char **b = b_start;

  while (a != a_end && b != b_end) {
    if (strcmp(*a, *b) < 0) {
      target[i] = *a;
      a++;
    } else {
      target[i] = *b;
      b++;
    }
    i++;
  }

  while (a != a_end) {
    target[i] = *a;
    a++;
    i++;
  }
  while (b != b_end) {
    target[i] = *b;
    b++;
    i++;
  }
}

void merge_sort_parallel(const char **start, const char **end) {
  if ((end - start) <= 1) {
    return;
  }

  if ((end - start) < CONCURRENT_THRESHOLD) {
    merge_sort_concurrent(start, end);
    return;
  }

  size_t half = (end - start) / 2;

  vector<const char *> left(start, start + half);
  vector<const char *> right(start + half, end);

  std::thread left_thread(
      [&left] { merge_sort_parallel(left.data(), left.data() + left.size()); });
  merge_sort_parallel(right.data(), right.data() + right.size());
  left_thread.join();

  merge(start, left.data(), left.data() + left.size(), right.data(),
        right.data() + right.size());
}

void merge_sort_concurrent(const char **start, const char **end) {
  if ((end - start) <= 1) {
    return;
  }

  size_t half = (end - start) / 2;

  vector<const char *> left(start, start + half);
  vector<const char *> right(start + half, end);

  merge_sort_concurrent(left.data(), left.data() + left.size());
  merge_sort_concurrent(right.data(), right.data() + right.size());

  merge(start, left.data(), left.data() + left.size(), right.data(),
        right.data() + right.size());
}

int main(int argc, const char **argv) {
  string input = argv[1];
  string output = argv[2];

  string line;
  vector<const char *> lines;
  ifstream in(input);
  while (getline(in, line)) {
    char *str = new char[line.size() + 1];
    strncpy(str, line.data(), line.size());
    str[line.size()] = '\0';
    lines.push_back(str);
  }
  in.close();

  merge_sort_parallel(lines.data(), lines.data() + lines.size());

  ofstream out(output);
  for (const auto &str : lines) {
    out << str << '\n';
  }
  out.close();

  for (const char *str : lines) {
    delete[] str;
  }

  return 0;
}

File: Benchmarks/MergeSort/Python/optimized.py
import multiprocessing
import sys
from typing import List

CONCURRENT_THRESHOLD = 10_000


def merge_sort_concurrent(entries: List[str]) -> None:
    if len(entries) <= 1:
        return

    half = len(entries) // 2
    left = entries[:half]
    right = entries[half:]

    merge_sort_concurrent(left)
    merge_sort_concurrent(right)

    merge(left, right, entries)


def merge_sort_parallel(entries: List[str]) -> None:
    if len(entries) <= CONCURRENT_THRESHOLD:
        merge_sort_concurrent(entries)
        return

    half = len(entries) // 2
    left = entries[:half]
    right = entries[half:]

    leftProcess = multiprocessing.Process(target=merge_sort_parallel, args=(left))
    leftProcess.start()

    merge_sort_parallel(right)

    leftProcess.join()
    leftProcess.close()

    merge(left, right, entries)


def merge(a: List[str], b: List[str], target: List[str]) -> None:
    i = 0
    ia = 0
    ib = 0

    while ia < len(a) and ib < len(b):
        if a[ia] < b[ib]:
            target[i] = a[ia]
            i += 1
            ia += 1
        else:
            target[i] = b[ib]
            i += 1
            ib += 1

    while ia < len(a):
        target[i] = a[ia]
        i += 1
        ia += 1
    while ib < len(b):
        target[i] = b[ib]
        i += 1
        ib += 1


if __name__ == "__main__":
    args = sys.argv

    input = args[1]
    output = args[2]

    with open(input, "r") as file:
        lines = file.readlines()

    merge_sort_concurrent(lines)

    with open(output, "w") as file:
        file.writelines(lines)
